apps/prairielearn/python/zygote.py

```diff
@@ -84,6 +87,33 @@
 prairielearn.get_unit_registry()
 
 
+# We want to conditionally allow/block importing specific modules, namely `rpy2`.
+# This custom importer will allow us to do so, and throw a custom error message.
+#
+# While this won't prevent anything more complex than an `import` statement, it
+# will make it clear to the user that they're not allowed to use `rpy2`. If they
+# try to bypass the block, it's up to them to deal with the consequences.
+class ForbidModuleMetaPathFinder(MetaPathFinder):
+    def __init__(self):
+        self.forbidden_modules: set[str] = set()
+
+    def forbid_module(self, forbidden_module: str):
+        self.forbidden_modules.add(forbidden_module)
+
+    def reset_forbidden_modules(self):
+        self.forbidden_modules.clear()
+
+    def find_spec(
+        self, fullname: str, path: Sequence[str] | None, target: types.ModuleType = None
```

Comment:
Is target defaulted to `None` in the base class?

Reply:
If nothing else, it's how the base class is documented: https://docs.python.org/3/library/importlib.html#importlib.abc.MetaPathFinder.find_spec

From what I can see, the base class doesn't actually have an implementation of this function.

Reply:
As it turns out, type checking wasn't even enabled for the zygote. I enabled it in 04593ad. After that, the type checker complains if I don't have a default for `target`.